java/ql/src/experimental/semmle/code/java/security/DecompressionBomb.qll

```diff
@@ -0,0 +1,384 @@
+import java
+private import semmle.code.java.dataflow.TaintTracking
+
+module DecompressionBomb {
+  /**
+   * The Decompression bomb Sink
+   *
+   * Extend this class for creating new decompression bomb sinks
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * The Additional flow steps that help to create a dataflow or taint tracking query
+   *
+   * Extend this class for creating new additional taint steps
+   */
+  class AdditionalStep extends Unit {
+    abstract predicate step(DataFlow::Node n1, DataFlow::Node n2);
+  }
+
+  abstract class BombReadInputStreamCall extends MethodCall { }
+
+  private class ReadInputStreamQualifierSink extends DecompressionBomb::Sink {
+    ReadInputStreamQualifierSink() { this.asExpr() = any(BombReadInputStreamCall r).getQualifier() }
+  }
+}
+
+/**
+ * Providing Decompression sinks and additional taint steps for `org.xerial.snappy` package
+ */
+module XerialSnappy {
+  /**
+   * A type that is responsible for `SnappyInputStream` Class
+   */
+  class TypeInputStream extends RefType {
+    TypeInputStream() {
+      this.getASupertype*().hasQualifiedName("org.xerial.snappy", "SnappyInputStream")
+    }
+  }
+
+  /**
+   * The methods that read bytes and belong to `SnappyInputStream` Types
+   */
+  class ReadInputStreamCall extends DecompressionBomb::BombReadInputStreamCall {
+    ReadInputStreamCall() {
+      this.getReceiverType() instanceof TypeInputStream and
+      this.getCallee().hasName(["read", "readNBytes", "readAllBytes"])
+    }
+  }
+
+  /**
+   * Gets `n1` and `n2` which `SnappyInputStream n2 = new SnappyInputStream(n1)` or
+   * `n1.read(n2)`,
+   *  second one is added because of sanitizer, we want to compare return value of each `read` or similar method
+   *  that whether there is a flow to a comparison between total read of decompressed stream and a constant value
+   */
+  private class InputStreamAdditionalTaintStep extends DecompressionBomb::AdditionalStep {
+    override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+      exists(Call call |
```

Comment:
```suggestion
      exists(ConstructorCall call |
```
Currently you are matching any method on `SnappyInputStream`, whereas you only want to match constructors. Please make this change in other places that you are matching constructors as well.